sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/action/QueryChangeStreamAction.java

```
@@ -270,28 +288,45 @@ public ProcessContinuation run(
-          LOG.debug("[" + token + "] Unable to update the current watermark, partition NOT FOUND");
-        } else {
-          LOG.error("[" + token + "] Error updating the current watermark: " + e.getMessage(), e);
-        }
-      }
-    };
-  }
-
   private boolean isTimestampOutOfRange(SpannerException e) {
     return (e.getErrorCode() == ErrorCode.INVALID_ARGUMENT
             || e.getErrorCode() == ErrorCode.OUT_OF_RANGE)
         && e.getMessage() != null
         && e.getMessage().contains(OUT_OF_RANGE_ERROR_MESSAGE);
   }
+
+  private void waitForAllChildTokensStartedRunning(String partitionToken) {
+    HashSet<String> childTokens = partitionMetadataDao.getChildTokens(partitionToken);
+    final long numChildTokens = childTokens.size();
+    long numRunningChildTokens = 0;
+
+    for (String childToken : childTokens) {
+      LOG.debug("[" + partitionToken + "] has " + numChildTokens + " child tokens and " +
+          numRunningChildTokens + " running child tokens");
+      boolean childTokenIsRunning = false;
+      do {
+        String state = partitionMetadataDao.getPartitionState(childToken);
+        if (state != null &&
+            (state.equals(State.FINISHED.name()) || state.equals(State.RUNNING.name()))) {
+          childTokenIsRunning = true;
+        } else {
+          try {
+            long numChildTokensToWaitFor = numChildTokens - numRunningChildTokens;
+            LOG.debug("[" + partitionToken + "] is waiting for " + numChildTokensToWaitFor +
+                "child tokens to start running");
+            // Approximate waiting times:
+            // 1. 1 partition: wait 1 second.
+            // 2. 100 partitions: wait 2 seconds.
+            // 3. 1000 partitions: wait 11 seconds.
+            // 4. 10000 partitions: wait ~1.5 minutes.
+            long millisToSleep = 1000 + numChildTokensToWaitFor * 10;
+            Thread.sleep(millisToSleep);
+          } catch (InterruptedException e) {
+            System.out.println(e);
```

We are swallowing exceptions here, please don't do this. Either wrap the exception in a `RuntimeException` and throw it or take another action.

Please don't use `System.out.println` as well.